### DISCUSSION TITLE
Questions on DQM

### DISCUSSION THREAD

**Discussion by mdpunch** [cta-observatory/nectarchain#95] [Q&A] [answered]
Hi there,

I have got the DQM running, though using @vmarandon 's script to get the runs):

Some questions (for @hashkar or ...):
* When I run it, at the end it closes the terminal emulator it's running in, including all other consoles open there!  This isn't very convenient, so I think it's a bug.  Can it be fixed?
* It also opens a blank tab in my browser. What is this meant to do?
* It produces by default a mysql file.  What should I do with this?
* I can use the option to produce plots also, in which case it makes a directory `Plots` in the output directory.  However, it doesn't put the plots there, but in the main output directory.  This seems to be a minor bug.
* The plots look to me to be fine, but maybe there is something I'm missing (maybe I could present them on Monday, at the NectarCAM verification meeting, and we could discuss)?

Good Luck,
Michael.

**Comment by jlenain** [comment]
Hi @mdpunch !

> Some questions (for @hashkar or ...):

I can at least try to answer some your points, below.

>     * When I run it, at the end it closes the terminal emulator it's running in, including all other consoles open there!  This isn't very convenient, so I think it's a bug.  Can it be fixed?
Very weird indeed, it does not happen on my side (tested on a laptop, a few desktops, all running Ubuntu, and a couple of servers - including EGI nodes - on EL7-like systems). Do you have a log of what happened by any chance, e.g. directing the stderr and stdout to file(s) when running the DQM ? Also, what is your desktop environment (Gnome, KDE, XFCE, ...) ? And your terminal emulator (Gnome terminal, terminator, xterm, ...) ?

>     * It also opens a blank tab in my browser. What is this meant to do?
It may have been related to the previous mechanism to automatically download run files from the grid, which uses web cookies to authenticate to the NectarCAM Elog under the hood. With `v0.1.7`, it may not occur.

>     * It produces by default a mysql file.  What should I do with this?
The DQM is actually meant to optionally use the NectarCAM daily sqlite files as input, in order to process monitoring data. If not downloaded first, the DQM tries to read it anyway, which produces an empty file, safe to be deleted.

>     * I can use the option to produce plots also, in which case it makes a directory `Plots` in the output directory.  However, it doesn't put the plots there, but in the main output directory.  This seems to be a minor bug.
Yes, this is a minor bug -> @hashkar ?

>     * The plots look to me to be fine, but maybe there is something I'm missing (maybe I could present them on Monday, at the NectarCAM verification meeting, and we could discuss)?
Sure thing !

**Comment by mdpunch** [comment]
> Hi @mdpunch !
> 
> > Some questions (for @hashkar or ...):
> 
> I can at least try to answer some your points, below.
> 
> > ```
> > * When I run it, at the end it closes the terminal emulator it's running in, including all other consoles open there!  This isn't very convenient, so I think it's a bug.  Can it be fixed?
> > ```
> 
> Very weird indeed, it does not happen on my side (tested on a laptop, a few desktops, all running Ubuntu, and a couple of servers - including EGI nodes - on EL7-like systems). Do you have a log of what happened by any chance, e.g. directing the stderr and stdout to file(s) when running the DQM ? Also, what is your desktop environment (Gnome, KDE, XFCE, ...) ? And your terminal emulator (Gnome terminal, terminator, xterm, ...) ?

I'm running Kubuntu, so it's closing the "Konsole".

I've tried it now in "Terminator" and despite the name, that one doesn't close!

But still, the job finishes with 
```
[1]+  Killed                  python /scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py $NECTARCAMDATA $NECTARDIR -i Beamer_runs/2023/20230512/NectarCAM.Run4333.0000.fits.fz -p -n >> log_file 2>> err_file
```
There is nothing particularly strange in the `stderr`, which ends with:
```
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
```
My machine (two weeks old, with 32GB ram, after the last one was lifted at CdG), has the CPUs all starting to be occupied, then the fan goes strongly, then the process ends (producing some plots, as mentioned).

So, something is going haywire, but not sure what.

> 
> > ```
> > * It also opens a blank tab in my browser. What is this meant to do?
> > ```
> 
> It may have been related to the previous mechanism to automatically download run files from the grid, which uses web cookies to authenticate to the NectarCAM Elog under the hood. With `v0.1.7`, it may not occur.

I will try updating my dev version to  `v0.1.7`...

> 
> > ```
> > * It produces by default a mysql file.  What should I do with this?
> > ```
> 
> The DQM is actually meant to optionally use the NectarCAM daily sqlite files as input, in order to process monitoring data. If not downloaded first, the DQM tries to read it anyway, which produces an empty file, safe to be deleted.
> 

Right, it's an empty file.  Will ignore.

> > ```
> > * I can use the option to produce plots also, in which case it makes a directory `Plots` in the output directory.  However, it doesn't put the plots there, but in the main output directory.  This seems to be a minor bug.
> > ```
> 
> Yes, this is a minor bug -> @hashkar ?
> 
> > ```
> > * The plots look to me to be fine, but maybe there is something I'm missing (maybe I could present them on Monday, at the NectarCAM verification meeting, and we could discuss)?
> > ```
> 
> Sure thing !

There is also a fits file created, with:
![image](https://github.com/cta-observatory/nectarchain/assets/17853426/4886dbc7-4156-4b39-ba0c-d4d9656a1b86)

Are there some scripts to view interesting things in this, or I should make them?

**Reply by mdpunch** [reply]
Well anyway, here are the plots from `2023/20230512/NectarCAM.Run4333.0000.fits.fz` (Is there some way to get it to go through all the files of a run together; this runs has 6 parts).

Since it's a Muon-Beamer run, the few static but repeated  images of a few muons can be seen in many of the plots, even though they're averages. 

So, my impression is that this part of this run at least is fine to share, even though 4 modules seem to be missing.

To be discussed!
 
![PlotsNectarCAM_Run4333_Camera_Mean_HighGain](https://github.com/cta-observatory/nectarchain/assets/17853426/50062b55-0f0d-4930-8b83-51d61c001037)
![PlotsNectarCAM_Run4333_Camera_Mean_LowGain](https://github.com/cta-observatory/nectarchain/assets/17853426/929d0757-951f-4ad7-8d5b-e5c0c029af33)
![PlotsNectarCAM_Run4333_Pedestal_Mean_HighGain](https://github.com/cta-observatory/nectarchain/assets/17853426/8c548789-d104-4a4f-abe6-94e35f3e3e85)
![PlotsNectarCAM_Run4333_Pedestal_Mean_LowGain](https://github.com/cta-observatory/nectarchain/assets/17853426/47dd9ed7-c411-482c-9ca4-ac2123e57d37)
![PlotsNectarCAM_Run4333_ChargeInt_Mean_HighGain_All](https://github.com/cta-observatory/nectarchain/assets/17853426/0059017f-6a4d-4a08-a485-f8166bccb497)
![PlotsNectarCAM_Run4333_ChargeInt_Mean_HighGain_Ped](https://github.com/cta-observatory/nectarchain/assets/17853426/df3f0546-471b-4e26-bb49-a397f98a0cd5)
![PlotsNectarCAM_Run4333_ChargeInt_Median_HighGain_All](https://github.com/cta-observatory/nectarchain/assets/17853426/96b62211-811a-4f5d-9238-da4d859941fc)
![PlotsNectarCAM_Run4333_ChargeInt_Median_HighGain_Ped](https://github.com/cta-observatory/nectarchain/assets/17853426/11d0df87-022c-4428-afcf-aa8b702d7e4c)
![PlotsNectarCAM_Run4333_ChargeInt_Rms_HighGain_All](https://github.com/cta-observatory/nectarchain/assets/17853426/32e958ee-b3ba-4f1b-a9d8-4df8d9be3927)
![PlotsNectarCAM_Run4333_ChargeInt_Rms_HighGain_Ped](https://github.com/cta-observatory/nectarchain/assets/17853426/780b4097-84aa-4236-9da0-5f26c81c720f)
![PlotsNectarCAM_Run4333_ChargeInt_Std_HighGain_All](https://github.com/cta-observatory/nectarchain/assets/17853426/181ee695-9263-4e1f-8e3c-af829f093cf7)
![PlotsNectarCAM_Run4333_ChargeInt_Std_HighGain_Ped](https://github.com/cta-observatory/nectarchain/assets/17853426/6384e250-e9a6-454d-aa92-1eac05de1328)
![PlotsNectarCAM_Run4333_Charge_Spectrum_HighGain_All](https://github.com/cta-observatory/nectarchain/assets/17853426/de44ad82-e810-459f-9e26-a1ef8ea9f90a)
![PlotsNectarCAM_Run4333_Charge_Spectrum_HighGain_Ped](https://github.com/cta-observatory/nectarchain/assets/17853426/a9820b1d-5b86-4262-9ec5-e176c7891166)

**Comment by jlenain** [comment]
> Are there some scripts to view interesting things in this, or I should make them?

Yes, there is a Bokeh app that I developed for that: https://github.com/cta-observatory/nectarchain/tree/master/src/nectarchain/dqm/bokeh_app, or did you mean something else ?

> Is there some way to get it to go through all the files of a run together; this runs has 6 parts

Yes, of course!
```shell
python /scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py --help
usage: start_dqm.py [-h] [-p] [--write-db] [-n] [-r RUNNB] [-i INPUT_FILES [INPUT_FILES ...]] input_paths output_paths

NectarCAM Data Quality Monitoring tool

positional arguments:
  input_paths           Input paths
  output_paths          Output paths

options:
  -h, --help            show this help message and exit
  -p, --plot            Enables plots to be generated
  --write-db            Write DQM output in DQM ZODB data base
  -n, --noped           Enables pedestal subtraction in charge integration
  -r RUNNB, --runnb RUNNB
                        Optional run number, automatically found on DIRAC
  -i INPUT_FILES [INPUT_FILES ...], --input-files INPUT_FILES [INPUT_FILES ...]
                        Local input files
```
so you can chain run files like that:
```shell
python /scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py $NECTARCAMDATA $NECTARDIR -i Beamer_runs/2023/20230512/NectarCAM.Run4333.0000.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0001.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0002.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0003.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0004.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0005.fits.fz -p -n >> log_file 2>> err_file
```
or just (easier):
```shell
python /scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py $NECTARCAMDATA $NECTARDIR -r 4333 -p -n >> log_file 2>> err_file
```

**Comment by mdpunch** [comment]
> > Are there some scripts to view interesting things in this, or I should make them?
> 
> Yes, there is a Bokeh app that I developed for that: https://github.com/cta-observatory/nectarchain/tree/master/src/nectarchain/dqm/bokeh_app, or did you mean something else ?

I'll check that out, thanks!
> 
> > Is there some way to get it to go through all the files of a run together; this runs has 6 parts
> 
> Yes, of course!
> 
...
> 
> so you can chain run files like that:
> 
> ```shell
> python /scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py $NECTARCAMDATA $NECTARDIR -i Beamer_runs/2023/20230512/NectarCAM.Run4333.0000.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0001.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0002.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0003.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0004.fits.fz Beamer_runs/2023/20230512/NectarCAM.Run4333.0005.fits.fz -p -n >> log_file 2>> err_file
> ```
I guess I meant if we could have a `glob` in there, to use ranges, wildcards, etc (so like `NectarCAM.Run4333.*.fits.fz`)

**But, running this with all the parts of the run lets the process go to completion without getting killed!**

So, DQM has a problem if only given part of a file?

> 
> or just (easier):
> 
> ```shell
> python /scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py $NECTARCAMDATA $NECTARDIR -r 4333 -p -n >> log_file 2>> err_file
> ```

That is indeed easier, and allows me to check out the downloading.  Unfortunately, in the `err_file` this gives me a series of:
```python
2024-01-26 13:39:34,869 nectarchain.data.management WARNING run 4333 is not present in ./runs/
NoneType: None
Traceback (most recent call last):
  File "/scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py", line 56, in <module>
    _, filelist = dm.findrun(args.runnb)
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/scr/punch/CTA/nectarchain/src/nectarchain/data/management.py", line 43, in findrun
    lfns = DataManagement.get_GRID_location(run_number)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scr/punch/CTA/nectarchain/src/nectarchain/data/management.py", line 121, in get_GRID_location
    return __class__.__get_GRID_location_DIRAC(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scr/punch/CTA/nectarchain/src/nectarchain/data/management.py", line 145, in __get_GRID_location_DIRAC
    fccli.do_find("-q " + basepath)
  File "/home/punch/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/DIRAC/DataManagementSystem/Client/FileCatalogClientCLI.py", line 1965, in do_find
    result = self.fc.findFilesByMetadata(metaDict, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/punch/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/DIRAC/Resources/Catalog/FileCatalog.py", line 158, in __getattr__
    raise AttributeError
AttributeError
```

I am in nectarchain  __version__  '0.1.7', and if I go into ipython, then after 
```python
from nectarchain.data.management import DataManagement
DataManagement.findrun(4333)
```
... I get the same errors as above

```python
In [2]: DataManagement.findrun(4333)
2024-01-26 14:13:32,099 nectarchain.data.management WARNING run 4333 is not present in ./runs/
NoneType: None
2024-01-26 13:13:32 UTC Framework/FileCatalog ERROR: FileCatalog._getEligibleCatalogs: Failed to get file catalog configuration. Path /Resources/FileCatalogs does not exist or it's not a section
2024-01-26 13:13:32 UTC Framework/FileCatalog ERROR: Failed to create catalog objects
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[2], line 1
----> 1 DataManagement.findrun(4333)

File /scr/punch/CTA/nectarchain/src/nectarchain/data/management.py:43, in DataManagement.findrun(run_number, search_on_GRID)
     41 log.warning(e, exc_info=True)
     42 log.info("will search files on GRID and fetch them")
---> 43 lfns = DataManagement.get_GRID_location(run_number)
     44 DataManagement.getRunFromDIRAC(lfns)
     45 list = glob.glob(
     46     basepath + "**/*" + str(run_number) + "*.fits.fz", recursive=True
     47 )

File /scr/punch/CTA/nectarchain/src/nectarchain/data/management.py:121, in DataManagement.get_GRID_location(run_number, output_lfns, basepath, fromElog, username, password)
    114     return __class__.__get_GRID_location_ELog(
    115         run_number=run_number,
    116         output_lfns=output_lfns,
    117         username=username,
    118         password=password,
    119     )
    120 else:
--> 121     return __class__.__get_GRID_location_DIRAC(
    122         run_number=run_number, basepath=basepath
    123     )

File /scr/punch/CTA/nectarchain/src/nectarchain/data/management.py:145, in DataManagement.__get_GRID_location_DIRAC(run_number, basepath)
    143 fccli = FileCatalogClientCLI(catalog.catalog)
    144 sys.stdout = StdoutRecord(keyword=f"Run{run_number}")
--> 145 fccli.do_find("-q " + basepath)
    146 lfns = sys.stdout.output
    147 sys.stdout = sys.__stdout__

File ~/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/DIRAC/DataManagementSystem/Client/FileCatalogClientCLI.py:1965, in FileCatalogClientCLI.do_find(self, args)
   1962 if verbose:
   1963     print("Query:", metaDict)
-> 1965 result = self.fc.findFilesByMetadata(metaDict, path)
   1966 if not result["OK"]:
   1967     print(f"Error: {result['Message']}")

File ~/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/DIRAC/Resources/Catalog/FileCatalog.py:158, in FileCatalog.__getattr__(self, name)
    156     return self.r_execute
    157 else:
--> 158     raise AttributeError

AttributeError: 
```
**... but if I run it a second time, it works!**

```python
In [3]: DataManagement.findrun(4333)
2024-01-26 13:13:58,557 nectarchain.data.management WARNING run 4333 is not present in ./runs/
NoneType: None
/vo.cta.in2p3.fr/nectarcam/2023/20230512/NectarCAM.Run4333.0000.fits.fz
/vo.cta.in2p3.fr/nectarcam/2023/20230512/NectarCAM.Run4333.0001.fits.fz
/vo.cta.in2p3.fr/nectarcam/2023/20230512/NectarCAM.Run4333.0002.fits.fz
/vo.cta.in2p3.fr/nectarcam/2023/20230512/NectarCAM.Run4333.0003.fits.fz
/vo.cta.in2p3.fr/nectarcam/2023/20230512/NectarCAM.Run4333.0004.fits.fz
/vo.cta.in2p3.fr/nectarcam/2023/20230512/NectarCAM.Run4333.0005.fits.fz
2024-01-26 13:14:01 UTC Framework/GFAL2_HTTPSStorage/GFAL2_StorageBase._getSingleFile INFO: Trying to download https://eos.grif.fr:11000/eos/grif/cta/vo.cta.in2p3.fr/nectarcam/2023/20230512/NectarCAM.Run4333.0000.fits.fz to /scr/punch/CTA/NectarCAM_muons/runs/NectarCAM.Run4333.0000.fits.fz
.0000.fits.fz
{'Failed': {},
 'Successful': {'/vo.cta.in2p3.fr/nectarcam/2023/20230512/NectarCAM.Run4333.0000.fits.fz': '/scr/punch/CTA/NectarCAM_muons/runs/NectarCAM.Run4333.0000.fits.fz'}}
2024-01-26 13:23:46 UTC Framework/GFAL2_HTTPSStorage/GFAL2_StorageBase._getSingleFile INFO: Trying to download https://eos.grif.fr:11000/eos/grif/cta/vo.cta.in2p3.fr/nectarcam/2023/20230512/NectarCAM.Run4333.0001.fits.fz to /scr/punch/CTA/NectarCAM_muons/runs/NectarCAM.Run4333.0001.fits.fz
...
```

**Reply by jlenain** [reply]
On your last point, yes, it's probably because you were too quick between the grid proxy initialization and the call to the DQM. By experience, you need to wait 10-20 sec after the proxy is initialized to be properly picked up by the DIRAC servers. But after that, you're good to go for 24h (the duration of a grid proxy).

**Comment by mdpunch** [comment]
> On your last point, yes, it's probably because you were too quick between the grid proxy initialization and the call to the DQM. By experience, you need to wait 10-20 sec after the proxy is initialized to be properly picked up by the DIRAC servers. But after that, you're good to go for 24h (the duration of a grid proxy).

I tried waiting 10 minutes, but still had to give the command twice in ipython!  Strange...

**Reply by jlenain** [reply]
Hi @mdpunch ,

Could you please try by applying this patch, and let me know if the connection with DIRAC works better ?
[management.txt](https://github.com/cta-observatory/nectarchain/files/14090174/management.txt)

**Reply by mdpunch** [reply]
Yes, now it is working correctly, in the `ipython` to get the file.  Thanks @jlenain !

What was the problem?  The `DIRAC.initialize()` helps how?

Now will try the full DQM.

**Reply by jlenain** [reply]
cf. https://dirac.readthedocs.io/en/latest/CodeDocumentation/Core/Base/Client.html#initialization

**Reply by mdpunch** [reply]
For the full DQM, running:
```shell
export NECTARDIR=./tmp
export NECTARCAMDATA="."
python /scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py $NECTARCAMDATA $NECTARDIR -r 4333 -p -n >> log_file 2>> err_file
```
... it downloads the runs, puts them in `./runs`, then stops with an error.
```shell
(nectar-dev) punch@apcnb2335:/scr/punch/CTA/NectarCAM_muons$ more err_file 
2024-01-30 08:35:45,278 nectarchain.data.management WARN run 4333 is not present in ./runs/
NoneType: None
Traceback (most recent call last):
  File "/scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py", line 109, in <module>
    reader = EventSource(input_url=path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/punch/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/ctapipe/io/eventsource.py", line 130, in __new__
    subcls = cls._find_compatible_source(input_url)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/punch/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/ctapipe/io/eventsource.py", line 354, in _find_compatible_source
    raise TraitError(
traitlets.traitlets.TraitError: input_url /scr/punch/CTA/NectarCAM_muons/NectarCAM.Run4333.0000.fits.fz is not an existing file  and no EventSource implementation claimed compatibility
(nectar-dev) punch@apcnb2335:/scr/punch/CTA/NectarCAM_muons$ ls ./runs
NectarCAM.Run4333.0000.fits.fz*  NectarCAM.Run4333.0002.fits.fz*  NectarCAM.Run4333.0004.fits.fz*
NectarCAM.Run4333.0001.fits.fz*  NectarCAM.Run4333.0003.fits.fz*  NectarCAM.Run4333.0005.fits.fz*
```

So somehow the DQM doesn't know how to load the runs it just got!  Strange...

**Reply by jlenain** [reply]
See #96 and #97

**Reply by mdpunch** [reply]
Downloading, running DQM starts okay.

But now, it doesn't produce the plots!

So, line 179 in start_dqm.py also needs to be changed from
```python
    path2 = f"{NectarPath}/{arg}"
```
to
```python
    path2 = f"{NectarPath}/runs/{arg}"
```

... and event then, it crashes before producing plots, with:
```python
Traceback (most recent call last):
  File "/scr/punch/CTA/nectarchain/src/nectarchain/dqm/start_dqm.py", line 185, in <module>
    for i, evt in enumerate(reader):
  File "/home/punch/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/ctapipe/io/eventsource.py", line 323, in __iter__
    for event in self._generator():
  File "/home/punch/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/ctapipe_io_nectarcam/__init__.py", line 374, in _generator
    self.fill_r0r1_container(array_event, event)
  File "/home/punch/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/ctapipe_io_nectarcam/__init__.py", line 691, in fill_r0r1_container
    r0, r1 = self.fill_r0r1_camera_container(zfits_event)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/punch/miniconda3/envs/nectar-dev/lib/python3.11/site-packages/ctapipe_io_nectarcam/__init__.py", line 677, in fill_r0r1_camera_container
    reshaped_waveform = zfits_event.waveform.reshape(N_GAINS, n_pixels, n_samples)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: cannot reshape array of size 110040 into shape (2,1834,60)
```
(I note, 110040/60 = 1834, so it looks like a problem with gain selection being applied but then not taken into account?)

**Reply by jlenain** [reply]
There are a couple of things to catch up: see #64 .

This boils down to disabling R0->R1 corrections, see the discussion here: https://github.com/cta-observatory/ctapipe_io_nectarcam/issues/34#issuecomment-1408984466

Very practically, you will need to apply the following patch in `ctapipe_io_nectarcam`:
```shell
diff --git a/src/ctapipe_io_nectarcam/calibration.py b/src/ctapipe_io_nectarcam/calibration.py
index 0e2d276..46febef 100644
--- a/src/ctapipe_io_nectarcam/calibration.py
+++ b/src/ctapipe_io_nectarcam/calibration.py
@@ -48,7 +48,7 @@ class NectarCAMR0Corrections(TelescopeComponent):
     ).tag(config=True)
 
     select_gain = Bool(
-        default_value=False,
+        default_value=True,
         help='Set to False to keep both gains.'
     ).tag(config=True)
```

**Reply by vmarandon** [reply]
Side remark : This type of error is potentially because of a bug in the EventBuilder (the version we are using at the moment) in which sometimes we have a "default" event that is written on disk and have only one gain and should not be there in this case : https://redmine.cta-observatory.org/issues/52457

**Reply by mdpunch** [reply]
I did the modification above, and it is running.

The problem is that it is running now for an hour... so lots of time for a nice quiche Lorraine, coffee, etc... hopefully it will finish some time this afternoon (but some kind of progress meter would be nice).

**Reply by jlenain** [reply]
Strongly agree, see https://github.com/cta-observatory/nectarchain/pull/81#issuecomment-1852430495

I will leave it to @hashkar to comment.

**Reply by hashkar** [reply]
We can discuss the DQM issue if you want during our next software meeting. I suspect  the charge spectrum plots are taking a long time to compute. I suggest only including them if necessary. 

@mdpunch how many events is there in your runs? Typically it takes me ~5 minutes to process ~10 000 events on my local machine

**Reply by jlenain** [reply]
Hi @hashkar , @mdpunch ,

The run we are talking about (run 4333) has ~10000 events per slice and 6 slices.

I tried to benchmark things a bit. With `nectarchain` v0.1.5, events are looped in the DQM and processed at a rate of ~75 events/second. In `nectarchain` v0.1.7, the rate drops to 1.35 events/second.

Taking a look at the differences in the DQM between these two versions (`git diff -u v0.1.5 v0.1.7 src/nectarchain/dqm`), I do not see any immediate culprit explaining such a difference.

`nectarchain` v0.1.5 depends on `ctapipe_io_nectarcam` v0.1.3, while with v0.1.5 I am now under `ctapipe_io_nectarcam` version 0.1.4.dev23+gb678680. The major difference there is the on-the-fly identification of trigger patches. @mdpunch , would it be possible that this introduced a major slow-down in the `EventSource` ?

**Reply by jlenain** [reply]
No luck, disabling the trigger patch identification in `ctapipe_io_nectarcam`:

```python
diff --git a/src/ctapipe_io_nectarcam/__init__.py b/src/ctapipe_io_nectarcam/__init__.py
index 6db0544..f81daf5 100644
--- a/src/ctapipe_io_nectarcam/__init__.py
+++ b/src/ctapipe_io_nectarcam/__init__.py
@@ -357,7 +357,7 @@ def load_camera_geometry(version=3):
     geom.frame = CameraFrame(focal_length=OPTICS.equivalent_focal_length)
 
     # Add the trigger patches as an attribute on-the-fly for now.
-    add_nectar_trigger_patches_to_geom(geom)    
+    # add_nectar_trigger_patches_to_geom(geom)
     
     return geom
```

does not change the event rate when reading a file.

**Comment by mdpunch** [comment]
Hi @jlenain, @hashkar,

Concerning the on-the-fly identification of trigger patches, I had pretty much optimized this I thought.  Then @maxnoe came along for the review, and made it 1500 times faster! (see https://github.com/cta-observatory/ctapipe_io_nectarcam/pull/40#discussion_r1238106902).

So, I don't think it's this, but maybe we could comment out that part, and compare the timing?

I'm not sure how to run a profiler on this, but that would be the ideal...

(I see you've already tried that... my message was too slow... but good to know my contribution - as improved by Max - isn't the cause)

**Reply by mdpunch** [reply]
On my machine, running on just the first file of the run (so 10,000 events), that takes 2h15, with the pedestal subtraction and plots.  Will try now without plots.

Edit: without plots takes 3h16 ???

**Reply by jlenain** [reply]
To be digested, but here are some profiler outputs under v0.1.5:
![dqm_0 1 5](https://github.com/cta-observatory/nectarchain/assets/5399958/58f60d9b-723d-4b5a-829b-fc5b1e23cf9b)

and under v0.1.7:
![dqm_0 1 7](https://github.com/cta-observatory/nectarchain/assets/5399958/10491c8e-1fb4-4968-bd91-e63dc9b55d7a)

**Reply by mdpunch** [reply]
So, is it `convert_to_pe`, or do I misinterpret?

( ... and can you tell me what tools you use for profiling?)

**Reply by jlenain** [reply]
> So, is it `convert_to_pe`, or do I misinterpret?

Yes, it seems so.

> ( ... and can you tell me what tools you use for profiling?)

In your conda environment, you need to additionally install `gprof2dot`.
For instance, for version `0.1.7`, I ran:
```shell
python -m cProfile -o dqm_0.1.7.pstats src/nectarchain/dqm/start_dqm.py -r 4333 -n $NECTARCAMDATA $NECTARCAMDATA
```
and then built the dependency graph with:
```shell
gprof2dot -f pstats dqm_v0.1.7.pstats | dot -Tpng -o dqm_0.1.7.png
```
For references: https://medium.com/@narenandu/profiling-and-visualization-tools-in-python-89a46f578989

**Reply by jlenain** [reply]
Here is a twist: I happened to reboot my laptop this morning, and came back to benchmarking the DQM.
I made several tests, always disabling the creation of plots:

- use the code as is
- disabling the R0->R1 correction altogether within the DQM
- using/not using the monitoring data

All tests end with an event rate of about 50 events/second, and the DQM is running quite fast, so I am a bit at a loss here...

@mdpunch , if you have the chance, could you please also test this after a reboot ?

**Reply by mdpunch** [reply]
Thanks.

Also, the calibration doesn't seem to work!  After running `calibrator(event)` on an event, the image looks like this:
```python
event.dl1.tel[0].image

array([1837.8861, 1858.7257,    0.    , ..., 1792.0222, 1872.7296,
       1882.7063], dtype=float32)
```
(this is whether `event.nectarcam.tel[0].evt.ucts_trigger_type` is 1 or 32).

I would have expected the pedestals to be subtracted after calibration, no?

I will try after reboot.

**Reply by mdpunch** [reply]
No luck after reboot... still running after 15 minutes on the first file of the run.

Also, on the pedestal subtraction, this is maybe normal that there is none!
**Edit:** No, not normal.

Looking at the pedestal-unsubtracted images:
![image](https://github.com/cta-observatory/nectarchain/assets/17853426/2d0309a5-e2ac-41d6-81f5-52a025970220)
... I realized from the image it's a "Cardboard Muon" run, for which there are no pedestals.
**Edit:** After checking with Vincent, there are (pedestals are `event.nectarcam.tel[0].evt.ucts_trigger_type==32`, and the camera triggers are type `1`).  So, still mysterious why they're not subtracted.


So, switching my testing to **Run 4741** ("Muon Beamer" run where Vincent said there are pedestals also)
**Edit:** This run unfortunately seems to not be on DIRAC yet.

**Reply by jlenain** [reply]
This is another question. The pedestal subtraction can be controlled with the `-p` option of the DQM. I think in all your examples, you did not activate the option.

**Reply by jlenain** [reply]
> So, switching my testing to Run 4741 ("Muon Beamer" run where Vincent said there are pedestals also)
> Edit: This run unfortunately seems to not be on DIRAC yet.

Yes it is, since Oct. 28, 2023:
```shell
dfind /vo.cta.in2p3.fr/nectarcam | grep 4741
/vo.cta.in2p3.fr/nectarcam/2023/20231025/NectarCAM.Run4741.0000.fits.fz
/vo.cta.in2p3.fr/nectarcam/2023/20231025/NectarCAM.Run4741.0001.fits.fz
/vo.cta.in2p3.fr/nectarcam/2023/20231025/NectarCAM.Run4741.0002.fits.fz
/vo.cta.in2p3.fr/nectarcam/2023/20231025/NectarCAM.Run4741.0003.fits.fz
```
What problem did you encounter ? Expired proxy, maybe ?

**Comment by jlenain** [comment, marked as answer]
Otherwise, I think I found the origin of the slow-down. I do not yet understand what the exact culprit is, but at least the following works:

- revert the change in `ctapipe_io_nectarcam` concerning gain selection, to be back to the HEAD
- on run 4333, everything runs normal if we limit to, say, 1000 events read per slice using the  `max_events` in `EventSource`
- otherwise, we hit the EventBuilder bug @vmarandon was refereeing to (https://redmine.cta-observatory.org/issues/52457)
- in such a case, this addition within the DQM should work:
[dqm_patch.txt](https://github.com/cta-observatory/nectarchain/files/14111686/dqm_patch.txt)

**Reply by mdpunch** [reply]
That is working much faster now (though I had to do part of the patch by hand, as I had other changes).

So, I've marked as answered.

Thanks @jlenain !

**Comment by mdpunch** [comment]
> This is another question. The pedestal subtraction can be controlled with the `-p` option of the DQM. I think in all your examples, you did not activate the option.

Here I'm also looking directly in my own script.  But you're correct that it's another question, so I'll open another thread!
(That way, I can describe more exactly).

**Comment by mdpunch** [comment]
> > So, switching my testing to Run 4741 ("Muon Beamer" run where Vincent said there are pedestals also)
> > Edit: This run unfortunately seems to not be on DIRAC yet.
> 
> Yes it is, since Oct. 28, 2023:
> 
> ```shell
> dfind /vo.cta.in2p3.fr/nectarcam | grep 4741
> /vo.cta.in2p3.fr/nectarcam/2023/20231025/NectarCAM.Run4741.0000.fits.fz
> /vo.cta.in2p3.fr/nectarcam/2023/20231025/NectarCAM.Run4741.0001.fits.fz
> /vo.cta.in2p3.fr/nectarcam/2023/20231025/NectarCAM.Run4741.0002.fits.fz
> /vo.cta.in2p3.fr/nectarcam/2023/20231025/NectarCAM.Run4741.0003.fits.fz
> ```
> 
> What problem did you encounter ? Expired proxy, maybe ?

Maybe, it's working now.

**Comment by mdpunch** [comment]
Closed since @jlenain found how to speed it up again, but there is still the bug that it closes the terminal emulator `konsole` when it finishes (but not if using the terminal emulator `terminator`), and still doesn't put the plots in the `Plots` directory it creates.
